File: lib/pwn/plugins/sandbox/driver.py
"""Disposable sandbox controller and in-isolation worker (stdlib only)."""
import base64
import hashlib
import json
import os
import pathlib
import random
import re
import resource
import shutil
import signal
import subprocess
import sys
import tempfile
import time
import uuid


def bounded(command, timeout, stdin=b'', memory=None):
    def limits():
        if memory:
            resource.setrlimit(resource.RLIMIT_AS, (memory * 1024 * 1024,) * 2)
        resource.setrlimit(resource.RLIMIT_FSIZE, (8 * 1024 * 1024,) * 2)
        resource.setrlimit(resource.RLIMIT_CORE, (0, 0))
    with tempfile.TemporaryFile() as out, tempfile.TemporaryFile() as err:
        process = subprocess.Popen(command, stdin=subprocess.PIPE, stdout=out, stderr=err, start_new_session=True, preexec_fn=limits)
        expired = False
        try:
            process.communicate(stdin, timeout=timeout)
        except subprocess.TimeoutExpired:
            expired = True
            os.killpg(process.pid, signal.SIGKILL)
            process.communicate()
        out.seek(0)
        err.seek(0)
        return {'exit': process.returncode, 'stdout': out.read(65536).decode(errors='replace'), 'stderr': err.read(65536).decode(errors='replace'), 'timed_out': expired}


def inside(request):
    command = ['/artifacts/target'] + request.get('argv', [])
    data = base64.b64decode(request['input'])
    timeout = request['timeout']
    memory = request['memory_mb']
    result = bounded(command, timeout, data, memory)
    sig = -result['exit'] if result['exit'] < 0 else None
    result.update(ok=True, signal=signal.Signals(sig).name if sig else None, faulting_address=None, backtrace=[], exploitability='unknown')
    # Instrumentation is a separate replay, explicitly not the original run.
    if shutil.which('strace'):
        trace = bounded(['strace', '-f', '-qq', '-s', '128', '--'] + command, timeout, data, memory)
        result['strace'] = trace['stderr']
    else:
        result['strace_error'] = 'strace unavailable in sandbox image'
    if sig and shutil.which('gdb'):
        pathlib.Path('/tmp/pwn-input').write_bytes(data)
        # GDB's run uses a shell: quote every argument; --args alone is not sufficient.
        import shlex
        run = 'run ' + ' '.join(shlex.quote(arg) for arg in command[1:]) + ' < /tmp/pwn-input'
        gdb = bounded(['gdb', '-q', '-nx', '-nh', '-batch', '-iex', 'set auto-load off', '-ex', run,
                       '-ex', 'p/x $pc', '-ex', 'bt', '-ex', 'exploitable', '--args', command[0]], timeout, memory=memory)
        text = gdb['stdout'] + gdb['stderr']
        result['gdb'] = text
        match = re.search(r'\$\d+ = (0x[0-9a-f]+)', text)
        result['faulting_address'] = match.group(1) if match else None
        result['backtrace'] = [line for line in text.splitlines() if re.match(r'#\d+\s', line)]
        classification = re.search(r'Exploitability Classification:\s*(\S+)', text)
        result['exploitability'] = classification.group(1) if classification else 'unknown: GDB exploitable plugin unavailable or failed'
    elif sig:
        result['gdb_error'] = 'gdb unavailable in sandbox image'
    return result


def snapshot(request):
    root = pathlib.Path.home() / '.pwn' / 'sandbox_snapshots'
    root.mkdir(parents=True, exist_ok=True, mode=0o700)
    target = pathlib.Path(request['binary']).resolve(strict=True)
    if not target.is_file():
        raise ValueError('binary must be a regular file')
    data = target.read_bytes()
    digest = hashlib.sha256(data).hexdigest()
    dest = root / uuid.uuid4().hex
    dest.mkdir(mode=0o700)
    (dest / 'target').write_bytes(data)
    (dest / 'target').chmod(0o500)
    (dest / 'manifest.json').write_text(json.dumps({'sha256': digest}))
    return {'ok': True, 'snapshot': str(dest), 'sha256': digest, 'semantics': 'immutable input snapshot; rollback creates a fresh disposable environment, not a live process checkpoint'}


def execute(request):
    backend = request.get('backend', 'docker')
    if backend not in ('docker', 'bwrap') or not shutil.which(backend):
        raise ValueError('sandbox backend unavailable: ' + backend)
    if backend == 'docker':
        probe = bounded(['docker', 'info', '--format', '{{.ServerVersion}}'], 10)
        if probe['exit']:
            raise ValueError('docker backend unavailable: ' + probe['stderr'])
    seconds = float(request.get('timeout', 10))
    memory = int(request.get('memory_mb', 256))
    if not 0 < seconds <= 300 or not 32 <= memory <= 4096:
        raise ValueError('timeout or memory budget out of range')
    argv = request.get('argv', [])
    if not isinstance(argv, list) or not all(isinstance(arg, str) and '\0' not in arg for arg in argv):
        raise ValueError('argv must be an array of strings without NUL')
    binary = pathlib.Path(request['binary']).resolve(strict=True)
    if not binary.is_file():
        raise ValueError('binary must be a regular file')
    with tempfile.TemporaryDirectory(prefix='pwn-sandbox-') as temp:
        artifacts = pathlib.Path(temp)
        shutil.copyfile(binary, artifacts / 'target')
        (artifacts / 'target').chmod(0o555)
        shutil.copyfile(__file__, artifacts / 'worker.py')
        payload = dict(request, timeout=seconds, memory_mb=memory, input=request.get('stdin_base64', base64.b64encode(request.get('stdin', '').encode()).decode()))
        container = 'pwn-sandbox-' + uuid.uuid4().hex
        if backend == 'docker':
            image = request.get('image', 'pwn-sandbox:local')
            if not re.fullmatch(r'[a-zA-Z0-9][a-zA-Z0-9_./:@-]*', image):
                raise ValueError('invalid image')
            cmd = ['docker', 'run', '--rm', '--pull=never', '--name', container, '-i', '--network=none', '--read-only', '--cap-drop=ALL',
                   '--security-opt=no-new-privileges', '--pids-limit=64', '--memory', str(memory)+'m', '--memory-swap', str(memory)+'m',
                   '--cpus=1', '--user=65534:65534', '--tmpfs=/tmp:rw,nosuid,nodev,size=64m',
                   '--mount', 'type=bind,src='+str(artifacts)+',dst=/artifacts,readonly', '--entrypoint=/usr/bin/python3', image, '-I', '/artifacts/worker.py', '--inside']
            artifacts.chmod(0o755)
        else:
            cmd = ['bwrap', '--unshare-all', '--die-with-parent', '--new-session', '--cap-drop', 'ALL', '--clearenv', '--setenv', 'PATH', '/usr/bin:/bin',
                   '--ro-bind', '/usr', '/usr']
            for name in ('/lib', '/lib64', '/bin', '/sbin'):
                if pathlib.Path(name).exists():
                    cmd += ['--ro-bind', name, name]
            cmd += ['--proc', '/proc', '--dev', '/dev', '--tmpfs', '/tmp', '--ro-bind', str(artifacts), '/artifacts', '--chdir', '/tmp', '/usr/bin/python3', '-I', '/artifacts/worker.py', '--inside']
        try:
            result = bounded(cmd, seconds * 3 + 10, json.dumps(payload).encode())
        finally:
            if backend == 'docker':
                bounded(['docker', 'rm', '-f', container], 10)
        if result['exit'] or result['timed_out']:
            raise ValueError('sandbox backend failed: ' + result['stderr'])
        response = json.loads(result['stdout'])
        response.update(backend=backend, network='none', artifact_mount='ro', memory_mb=memory, timeout=seconds,
                        isolation_limitations='bwrap uses per-process RLIMIT_AS, not aggregate cgroup memory/pid limits' if backend == 'bwrap' else None)
        return response


def main(request):
    action = request.get('action', 'run')
    if action == 'snapshot':
        return snapshot(request)
    if action == 'rollback':
        path = pathlib.Path(request['snapshot']).resolve(strict=True)
        root = (pathlib.Path.home() / '.pwn' / 'sandbox_snapshots').resolve()
        if path.parent != root:
            raise ValueError('unknown snapshot')
        expected = json.loads((path / 'manifest.json').read_text())['sha256']
        if hashlib.sha256((path / 'target').read_bytes()).hexdigest() != expected:
            raise ValueError('snapshot integrity failure')
        return execute(dict(request, binary=str(path / 'target')))
    if action == 'fuzz':
        minutes = float(request.get('minutes', 1))
        if not 0 < minutes <= 60:
            raise ValueError('minutes must be 0..60')
        corpus = pathlib.Path(request['corpus']).resolve(strict=True)
        seeds = [p.read_bytes()[:65536] for p in sorted(corpus.iterdir()) if p.is_file() and not p.is_symlink()][:256]
        if not seeds:
            raise ValueError('empty corpus')
        rng = random.Random(request.get('seed', 0))
        deadline = time.monotonic() + minutes * 60
        crashes = []
        iterations = 0
        while time.monotonic() < deadline:
            data = bytearray(rng.choice(seeds) or b'\0')
            data[rng.randrange(len(data))] ^= 1 << rng.randrange(8)
            # Preserve byte-exact mutated inputs across the JSON transport.
            result = execute(dict(request, stdin_base64=base64.b64encode(data).decode(), timeout=min(float(request.get('timeout', 2)), max(.01, (deadline-time.monotonic())/3))))
            iterations += 1
            if result.get('signal'):
                crashes.append(dict(result, input_base64=base64.b64encode(data).decode()))
                if len(crashes) >= 16:
                    break
        return {'ok': True, 'backend': request.get('backend', 'docker'), 'iterations': iterations, 'crashes': crashes, 'seed': request.get('seed', 0)}
    return execute(request)


if __name__ == '__main__':
    try:
        request = json.load(sys.stdin)
        response = inside(request) if '--inside' in sys.argv else main(request)
    except Exception as error:
        response = {'ok': False, 'error': str(error)}
    print(json.dumps(response))

File: lib/pwn/plugins/sandbox/driver.rb
# frozen_string_literal: true

require 'base64'
require 'digest'
require 'fileutils'
require 'json'
require 'securerandom'
require 'shellwords'
require 'tempfile'
require 'tmpdir'

# Standalone stdlib controller; target execution occurs only in the worker.
module PWNSandboxDriver
  public_class_method def self.authors
    'AUTHOR(S): 0day Inc. <support@0dayinc.com>'
  end

  public_class_method def self.help
    puts "Internal standalone Ruby sandbox controller/worker; use PWN::Plugins::Sandbox for the supported API.
      # Print the module author information.
      #{self}.authors"
  end

  class << self
    def available?(name)
      ENV.fetch('PATH', '').split(File::PATH_SEPARATOR).any? { |dir| File.executable?(File.join(dir, name)) }
    end

    def bounded(command, seconds, input = '', memory = nil)
      Tempfile.create('pwn-out') do |out|
        Tempfile.create('pwn-err') do |err|
          source, sink = IO.pipe
          begin
            source.binmode
            sink.binmode
            limits = { pgroup: true, in: source, out: out, err: err, rlimit_fsize: [8 * 1024 * 1024] * 2, rlimit_core: [0, 0] }
            limits[:rlimit_as] = [memory * 1024 * 1024] * 2 if memory
            pid = Process.spawn(*command, **limits)
            source.close
            writer = Thread.new do
              sink.write(input)
            rescue Errno::EPIPE, IOError
              nil
            ensure
              sink.close unless sink.closed?
            end
            deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + seconds
            expired = false
            status = nil
            loop do
              waited = Process.waitpid2(pid, Process::WNOHANG)
              if waited
                status = waited.last
                break
              end
              if Process.clock_gettime(Process::CLOCK_MONOTONIC) >= deadline
                expired = true
                Process.kill('KILL', -pid)
                status = Process.waitpid2(pid).last
                break
              end
              sleep 0.005
            end
            { exit: status.signaled? ? -status.termsig : status.exitstatus,
              stdout: File.binread(out.path, 65_536).to_s.dup.force_encoding('UTF-8').scrub,
              stderr: File.binread(err.path, 65_536).to_s.dup.force_encoding('UTF-8').scrub, timed_out: expired }
          ensure
            source.close unless source.closed?
            writer&.kill
            writer&.join(1)
            sink.close unless sink.closed?
            if pid
              begin
                Process.kill('KILL', -pid)
              rescue Errno::ESRCH
                nil
              end
              begin
                Process.waitpid(pid)
              rescue Errno::ECHILD
                nil
              end
            end
          end
        end
      end
    end

    def inside(request)
      command = ['/artifacts/target'] + request.fetch(:argv, [])
      data = Base64.strict_decode64(request.fetch(:input))
      seconds = request.fetch(:timeout)
      memory = request.fetch(:memory_mb)
      result = bounded(command, seconds, data, memory)
      sig = result[:exit].negative? ? Signal.signame(-result[:exit]) : nil
      result.merge!(ok: true, signal: sig && "SIG#{sig}", faulting_address: nil, backtrace: [], exploitability: 'unknown')
      if available?('strace')
        result[:strace] = bounded(['strace', '-f', '-qq', '-s', '128', '--'] + command, seconds, data, memory)[:stderr]
      else
        result[:strace_error] = 'strace unavailable in sandbox image'
      end
      if sig && available?('gdb')
        File.binwrite('/tmp/pwn-input', data)
        run = "run #{command.drop(1).shelljoin} < /tmp/pwn-input"
        replay = bounded(['gdb', '-q', '-nx', '-nh', '-batch', '-iex', 'set auto-load off', '-ex', run,
                          '-ex', 'p/x $pc', '-ex', 'bt', '-ex', 'exploitable', '--args', command.first], seconds, '', memory)
        text = replay[:stdout] + replay[:stderr]
        result[:gdb] = text
        result[:faulting_address] = text[/\$\d+ = (0x[0-9a-f]+)/, 1]
        result[:backtrace] = text.lines.map(&:chomp).grep(/\A#\d+\s/)
        result[:exploitability] = text[/Exploitability Classification:\s*(\S+)/, 1] || 'unknown: GDB exploitable plugin unavailable or failed'
      elsif sig
        result[:gdb_error] = 'gdb unavailable in sandbox image'
      end
      result
    end
  end

  # Host-side orchestration never launches a target outside a backend.
  class << self
    def execute(request)
      backend = request.fetch(:backend, 'docker')
      raise ArgumentError, "sandbox backend unavailable: #{backend}" unless %w[docker bwrap].include?(backend) && available?(backend)

      if backend == 'docker'
        probe = bounded(['docker', 'info', '--format', '{{.ServerVersion}}'], 10)
        raise ArgumentError, "docker backend unavailable: #{probe[:stderr]}" unless probe[:exit].zero?
      end
      seconds = Float(request.fetch(:timeout, 10))
      memory = Integer(request.fetch(:memory_mb, 256))
      raise ArgumentError, 'timeout or memory budget out of range' unless seconds.positive? && seconds <= 300 && memory.between?(32, 4096)

      argv = request.fetch(:argv, [])
      raise ArgumentError, 'argv must be an array of strings without NUL' unless argv.is_a?(Array) && argv.all? { |arg| arg.is_a?(String) && !arg.include?("\0") }

      binary = File.realpath(request.fetch(:binary))
      raise ArgumentError, 'binary must be a regular file' unless File.file?(binary)

      Dir.mktmpdir('pwn-sandbox-') do |artifacts|
        FileUtils.copy_file(binary, File.join(artifacts, 'target'))
        File.chmod(0o555, File.join(artifacts, 'target'))
        FileUtils.copy_file(__FILE__, File.join(artifacts, 'worker.rb'))
        payload = request.merge(timeout: seconds, memory_mb: memory, input: request.fetch(:stdin_base64) { Base64.strict_encode64(request.fetch(:stdin, '')) })
        container = "pwn-sandbox-#{SecureRandom.hex(16)}"
        if backend == 'docker'
          image = request.fetch(:image, 'pwn-sandbox:local')
          raise ArgumentError, 'invalid image' unless %r{\A[a-zA-Z0-9][a-zA-Z0-9_./:@-]*\z}.match?(image)

          File.chmod(0o755, artifacts)
          command = ['docker', 'run', '--rm', '--pull=never', '--name', container, '-i', '--network=none', '--read-only', '--cap-drop=ALL',
                     '--security-opt=no-new-privileges', '--pids-limit=64', '--memory', "#{memory}m", '--memory-swap', "#{memory}m",
                     '--cpus=1', '--user=65534:65534', '--tmpfs=/tmp:rw,nosuid,nodev,size=64m',
                     '--mount', "type=bind,src=#{artifacts},dst=/artifacts,readonly", '--entrypoint=/usr/bin/ruby', image, '/artifacts/worker.rb', '--inside']
        else
          command = ['bwrap', '--unshare-all', '--die-with-parent', '--new-session', '--cap-drop', 'ALL', '--clearenv', '--setenv', 'PATH', '/usr/bin:/bin', '--ro-bind', '/usr', '/usr']
          %w[/lib /lib64 /bin /sbin].each { |name| command.push('--ro-bind', name, name) if File.exist?(name) }
          command.push('--proc', '/proc', '--dev', '/dev', '--tmpfs', '/tmp', '--ro-bind', artifacts, '/artifacts', '--chdir', '/tmp', File.realpath('/usr/bin/ruby'), '/artifacts/worker.rb', '--inside')
        end
        begin
          result = bounded(command, (seconds * 3) + 10, JSON.generate(payload))
        ensure
          bounded(['docker', 'rm', '-f', container], 10) if backend == 'docker'
        end
        raise ArgumentError, "sandbox backend failed: #{result[:stderr]}" if !result[:exit].zero? || result[:timed_out]

        JSON.parse(result[:stdout], symbolize_names: true).merge(backend: backend, network: 'none', artifact_mount: 'ro', memory_mb: memory, timeout: seconds,
                                                                 isolation_limitations: backend == 'bwrap' ? 'bwrap uses per-process RLIMIT_AS, not aggregate cgroup memory/pid limits' : nil)
      end
    end

    def snapshot(request)
      root = File.join(Dir.home, '.pwn', 'sandbox_snapshots')
      FileUtils.mkdir_p(root, mode: 0o700)
      target = File.realpath(request.fetch(:binary))
      raise ArgumentError, 'binary must be a regular file' unless File.file?(target)

      data = File.binread(target)
      digest = Digest::SHA256.hexdigest(data)
      dest = File.join(root, SecureRandom.hex(16))
      Dir.mkdir(dest, 0o700)
      File.binwrite(File.join(dest, 'target'), data)
      File.chmod(0o500, File.join(dest, 'target'))
      File.write(File.join(dest, 'manifest.json'), JSON.generate(sha256: digest))
      { ok: true, snapshot: dest, sha256: digest, semantics: 'immutable input snapshot; rollback creates a fresh disposable environment, not a live process checkpoint' }
    end

    def fuzz(request)
      minutes = Float(request.fetch(:minutes, 1))
      raise ArgumentError, 'minutes must be 0..60' unless minutes.positive? && minutes <= 60

      corpus = File.realpath(request.fetch(:corpus))
      seeds = Dir.children(corpus).sort.filter_map do |name|
        path = File.join(corpus, name)
        File.binread(path, 65_536).to_s if File.file?(path) && !File.symlink?(path)
      end.take(256)
      raise ArgumentError, 'empty corpus' if seeds.empty?

      rng = Random.new(request.fetch(:seed, 0))
      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + (minutes * 60)
      crashes = []
      iterations = 0
      while Process.clock_gettime(Process::CLOCK_MONOTONIC) < deadline
        data = seeds.fetch(rng.rand(seeds.length)).dup
        data = +"\x00" if data.empty?
        index = rng.rand(data.bytesize)
        data.setbyte(index, data.getbyte(index) ^ (1 << rng.rand(8)))
        encoded = Base64.strict_encode64(data)
        remaining = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
        seconds = [Float(request.fetch(:timeout, 2)), (remaining / 3).clamp(0.01, Float::INFINITY)].min
        result = execute(request.merge(stdin_base64: encoded, timeout: seconds))
        return result unless result[:ok]

        iterations += 1
        crashes << result.merge(input_base64: encoded) if result[:signal]
        break if crashes.length >= 16
      end
      { ok: true, backend: request.fetch(:backend, 'docker'), iterations: iterations, crashes: crashes, seed: request.fetch(:seed, 0) }
    end

    def main(request)
      case request.fetch(:action, 'run')
      when 'fuzz'
        fuzz(request)
      when 'snapshot'
        snapshot(request)
      when 'rollback'
        path = File.realpath(request.fetch(:snapshot))
        root = File.realpath(File.join(Dir.home, '.pwn', 'sandbox_snapshots'))
        raise ArgumentError, 'unknown snapshot' unless File.dirname(path) == root

        target = File.join(path, 'target')
        expected = JSON.parse(File.read(File.join(path, 'manifest.json'))).fetch('sha256')
        raise ArgumentError, 'snapshot integrity failure' unless Digest::SHA256.file(target).hexdigest == expected

        execute(request.merge(binary: target))
      else
        execute(request)
      end
    end
  end
end

if $PROGRAM_NAME == __FILE__
  begin
    request = JSON.parse($stdin.read, symbolize_names: true)
    response = ARGV == ['--inside'] ? PWNSandboxDriver.inside(request) : PWNSandboxDriver.main(request)
  rescue StandardError => e
    response = { ok: false, error: e.message }
  end
  puts JSON.generate(response)
end
